codeforgreenville/c4g-events
# MattT/Add Healthcheck Route

Fixes #26

# Summary
Places a FastAPI instance out in front of the Bolt setup to allow for various additionals types of requests to be handled. Slack bot functionality is kept intact by utilizing an [adapter](https://slack.dev/bolt-python/concepts#adapters) to allow Bolt to communicate via FastAPI to the rest of the world. I only picked FastAPI because that's what I'm most familiar with -- any number of web frameworks can be used in its place and I wouldn't be offended if another one were decided to be preferable.

A `/healthz` "heartbeat" route has been added that will return a 500 if any threads are found to have died.

A [healthcheck](https://docs.docker.com/compose/compose-file/compose-file-v3/#healthcheck) section has been added to the docker-compose.yml file. If curl does not receive a successful response from the healthcheck route more than 3 times in a row then the container will be marked as "unhealthy"; at that point the `willfarrell/autoheal` container will know to mop up and force a restart of the slack-events-bot container.

I've also added a small pytest setup and a new workflow step to run automated test in the CI workflow.

# Screenshots
Forcing a restart by spamming a non-existent route:
![Screenshot_2023-09-28_21-42-39](https://github.com/hackgvl/slack-events-bot/assets/44626690/5466fb5b-e88a-433d-83db-1b97b3ed0223)


# Test Plan
1. I set the docker-compose file to build from a local context instead of pulling an image from Docker Hub. This could very well be an optional step if you have another way of going about it.
```  slack-events-bot:
    # image: hackgvl/slack-events-bot
    build:
      context: .
      dockerfile: Dockerfile
```
2. Change [this line](https://github.com/ThorntonMatthewD/slack-events-bot/blob/3e332b695e10166e99bf136df786000be6a0fc76/src/bot.py#L219) to the following in order to force the unhappy path:
```python
        if not False:
            ...
```

3. Run `docker-compose --profile autohealing up` to start both of the containers.
4. Run `docker ps -aqf "name=^slack-events-bot$"` and note the ID. 
5. Wait for the healthcheck to fail 3 times, and then wait for the autoheal container's check interval to roll around. Feel free to tweak the numbers in the docker-compose.yml file if you would like to expedite things.
6. You should see the following message:
```
... Container /slack-events-bot (<ID from step 4>) found to be unhealthy - Restarting container now with 10s timeout
```
7. Once the new container is up, rerun `docker ps -aqf "name=^slack-events-bot$"` and ensure that you have a different ID now.
